### ISSUE TITLE
Toggle inline / reference link.

### ISSUE THREAD

**Issue by MonsieurMan** [neilsustc/vscode-markdown#260]
It would be useful to have a refactoring that allows you to toggle between inline and reference link.  

**Before**
```md
[Example link](http://example.com)
```
**After**
```md
[Example][1]
[1]: http://example.com
```

> I do it manually all the time and it makes me quite sad 😢

**Comment by neilsustc** [comment]
I agree that it might be useful. But I probably have little time for this in the near future. Let's see whether there is somebody can help.

**Comment by MonsieurMan** [comment]
I don't have that much time the two next weeks neither but I could definitely help creating it after. 😃

**Comment by RoyiAvital** [comment]
Is there any progress on this?

**Comment by yzhang-gh** [comment]
Thanks for asking.

Technically, it is not hard (I guess). But there are many corner cases to take into account. It would be great if anyone can help as now I only have limited time for this extension (for crucial bugs or small fixes).

**Comment by RoyiAvital** [comment]
There is a reference `Python` code here: https://stackoverflow.com/a/30919298.
It is not perfect, I wish I knew enough `Python` to fix it.

But if you copy it to `JS` and implement the feature behind an opt in option I can test it and submit the edge cases.
Probably from there things will get better, right?

**Comment by yzhang-gh** [comment]
Well, if you look at the "Cases" section in the origin post of this issue, you will see a lot of implementation details, which are not considered in the link you mentioned. For example,

```markdown
<!-- From: -->
[My [simple] link](<uri uri> 'title \'n )(title')

<!-- To: -->
[My [simple] link][2_ref-def]

[2_ref-def]: <uri%20uri> "title 'n )(title"
```

note `\'` in the `title` field. Also,

> Inline links with the same destination and title should refer to the same reference link definition.

> Code spans and code blocks should remain untouched.

and more.

As I said, it is not technically hard, but requires a certain amount of time. It cannot be a one-pass regex replacement but needs a reasonable workflow (e.g. A rough thinking tells me it should have 3 (maybe 2) steps: parse existing reference link definitions, identify inline/reference links, and then transform). If anyone can make a PR, I'm happy to review it.
